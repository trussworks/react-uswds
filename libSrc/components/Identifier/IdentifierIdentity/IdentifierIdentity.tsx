import React from 'react'
import classnames from 'classnames'

export interface BaseIdentifierIdentityProps {
  children: React.ReactNode
  domain: string
  className?: string
}

export type IdentifierIdentityProps = BaseIdentifierIdentityProps &
  JSX.IntrinsicElements['div']

const IdentifierIdentity = ({
  children,
  domain,
  className,
  ...divProps
}: IdentifierIdentityProps): React.ReactElement => {
  const classes = classnames('usa-identifier__identity', className)
  return (
    <div data-testid="identifierIdentity" className={classes} {...divProps}>
      <p className="usa-identifier__identity-domain">{domain}</p>
      <p className="usa-identifier__identity-disclaimer">{children}</p>
    </div>
  )
}

export default IdentifierIdentity
