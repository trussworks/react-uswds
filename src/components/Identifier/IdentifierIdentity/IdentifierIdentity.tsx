import React from 'react'
import classnames from 'classnames'

interface IdentifierIdentityProps {
  children: React.ReactNode
  domain: string
  className?: string
}

export const IdentifierIdentity = ({
  children,
  domain,
  className,
  ...divProps
}: IdentifierIdentityProps &
  JSX.IntrinsicElements['div']): React.ReactElement => {
  const classes = classnames('usa-identifier__identity', className)
  return (
    <div data-testid="identifierIdentity" className={classes} {...divProps}>
      <p className="usa-identifier__identity-domain">{domain}</p>
      <p className="usa-identifier__identity-disclaimer">{children}</p>
    </div>
  )
}

export default IdentifierIdentity
