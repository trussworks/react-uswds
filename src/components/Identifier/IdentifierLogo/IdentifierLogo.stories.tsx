import React from 'react'

import { IdentifierLogo } from './IdentifierLogo'

// assets
// import logoImg from 'uswds/src/img/logo-img.png'
import dotGovIcon from 'uswds/src/img/icon-dot-gov.svg'

export default {
  title: 'Components/Identifier/Logo',
  component: IdentifierLogo,
  parameters: {
    docs: {
      description: {
        component: `
        Display optional agency logos as links.  Used in USWDS 2.0 Identifier component.
        
        Source: https://designsystem.digital.gov/components/identifier/
        `,
      },
    },
  },
}

export const OneLogo = (): React.ReactElement => (
  <div className="">
    <IdentifierLogo
      image={<img className="" src={dotGovIcon} alt="Mock logo" />}
    />
  </div>
)

export const TwoLogos = (): React.ReactElement => (
  <div className="">
    <IdentifierLogo
      image={<img className="" src={dotGovIcon} alt="Mock logo" />}
    />
    <IdentifierLogo
      image={<img className="" src={dotGovIcon} alt="Mock logo" />}
    />
  </div>
)
