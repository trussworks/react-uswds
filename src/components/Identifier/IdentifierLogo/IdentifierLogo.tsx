import React from 'react'
import classnames from 'classnames'

export interface IdentifierLogoProps {
  children: React.ReactNode
  className?: string
}

export const IdentifierLogo = ({
  children,
  className,
  ...anchorProps
}: IdentifierLogoProps & JSX.IntrinsicElements['a']): React.ReactElement => {
  const classes = classnames('usa-identifier__logo', className)
  return (
    <a className={classes} {...anchorProps}>
      {children}
    </a>
  )
}
