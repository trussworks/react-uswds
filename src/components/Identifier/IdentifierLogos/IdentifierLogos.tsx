import React from 'react'
import classnames from 'classnames'

interface IdentifierLogosProps {
  children: React.ReactNode
  className?: string
}

export const IdentifierLogos = ({
  children,
  className,
  ...divProps
}: IdentifierLogosProps & JSX.IntrinsicElements['div']): React.ReactElement => {
  const classes = classnames('usa-identifier__logos', className)
  return (
    <div data-testid="identifierLogos" className={classes} {...divProps}>
      {children}
    </div>
  )
}

export default IdentifierLogos
