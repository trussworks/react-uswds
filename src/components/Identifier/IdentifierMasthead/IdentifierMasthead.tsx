import React from 'react'
import classnames from 'classnames'

interface IdentifierMastheadProps {
  className?: string
  children?: React.ReactNode
}

export const IdentifierMasthead = ({
  className,
  children,
  ...sectionProps
}: IdentifierMastheadProps &
  JSX.IntrinsicElements['section']): React.ReactElement => {
  const classes = classnames(
    'usa-identifier__section usa-identifier__section--masthead',
    className
  )

  return (
    <section
      data-testid="identifierMasthead"
      className={classes}
      {...sectionProps}>
      <div className="usa-identifier__container">{children}</div>
    </section>
  )
}

export default IdentifierMasthead
