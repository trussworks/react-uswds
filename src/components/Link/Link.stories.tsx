import React from 'react'
import { Link } from './Link'

export default {
  title: 'Components/Typography/Link',
  component: Link,
  parameters: {
    docs: {
      description: {
        component: `
Based on Typography guide: https://designsystem.digital.gov/components/typography/#links
`,
      },
    },
  },
}

export const Default = (): React.ReactElement => (
  <p>
    <Link href={'#'}> This</Link> is a text link.
  </p>
)

export const Visited = (): React.ReactElement => (
  <p>
    <Link className="usa-color-text-visited" href={'#'}>
      This
    </Link>
    &nbsp;is a visited link.
  </p>
)

export const DarkBackground = (): React.ReactElement => (
  <div className="usa-dark-background" style={{ padding: '10px' }}>
    <p>
      <Link href={'#'}>This</Link> is a text link on a dark background
    </p>
  </div>
)

export const ExternalLink = (): React.ReactElement => (
  <p>
    This is a link that goes to an&nbsp;
    <Link variant="external" href={'http://www.truss.works'}>
      external website
    </Link>
    .
  </p>
)

export const NavLink = (): React.ReactElement => (
  <p>
    This is a&nbsp;
    <Link variant="nav" href={'#'}>
      NavLink
    </Link>
  </p>
)

export const StyledAsButton = (): React.ReactElement => (
  <p>
    <Link className="usa-button" variant="unstyled" href={'#'}>
      This
    </Link>
    is a link styled to look like a button.
  </p>
)

export const CustomComponentLink = (): React.ReactElement => {
  type MockLinkProps = React.PropsWithChildren<{
    to: string
    className: string
  }> &
    JSX.IntrinsicElements['a']

  const CustomLink: React.FunctionComponent<MockLinkProps> = ({
    to,
    className,
    children,
    ...linkProps
  }: MockLinkProps): React.ReactElement => (
    <a href={to} className={className} {...linkProps}>
      {children}
    </a>
  )

  return (
    <p>
      <Link<MockLinkProps>
        className="abc"
        asCustom={CustomLink}
        to="http://www.truss.works">
        This
      </Link>
      &nbsp;is a custom component link.
    </p>
  )
}
