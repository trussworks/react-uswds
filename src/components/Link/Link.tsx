import React from 'react'
import classnames from 'classnames'

// These props we want to require always, even on custom components
type StyledLinkProps<T> = {
  variant?: 'external' | 'unstyled' | 'nav'
  className?: string
  children: React.ReactNode
} & T

// These props are only required on the default Link
interface WithDefaultLinkProps {
  href: string
}

// Add `asCustom` to the provided custom props
interface WithCustomLinkProps<T> {
  asCustom: React.FunctionComponent<T>
}

// Default props means allow the StyledLinkProps as well as any
// props allowed on the `a` element, plus the required props on
// WithDefaultLinkProps
export type DefaultLinkProps = StyledLinkProps<JSX.IntrinsicElements['a']> &
  WithDefaultLinkProps

// Custom props means allow the StyledLinkProps as well as the custom
// props, plus the required props on WithCustomLinkProps
export type CustomLinkProps<T> = StyledLinkProps<T> & WithCustomLinkProps<T>

export function isCustomProps<T>(
  props: DefaultLinkProps | CustomLinkProps<T>
): props is CustomLinkProps<T> {
  return 'asCustom' in props
}

function linkClasses<T>(
  variant: StyledLinkProps<T>['variant'],
  className: StyledLinkProps<T>['className']
): string | undefined {
  const unstyled = variant === 'unstyled'
  const isExternalLink = variant === 'external'
  const isNavLink = variant === 'nav'

  return unstyled
    ? className
    : classnames(
        'usa-link',
        { 'usa-link--external': isExternalLink, 'usa-nav__link': isNavLink },
        className
      )
}

export function Link(props: DefaultLinkProps): React.ReactElement
export function Link<T>(props: CustomLinkProps<T>): React.ReactElement
export function Link<FCProps = DefaultLinkProps>(
  props: DefaultLinkProps | CustomLinkProps<FCProps>
): React.ReactElement {
  if (isCustomProps(props)) {
    const { variant, className, asCustom, children, ...remainingProps } = props
    // 1. We know props is AsCustomProps<FCProps>
    // 2. We know AsCustomProps<FCProps> is
    //    FCProps & { variant: ..., className: ..., children: ..., asCustom: ... }
    // 3. Therefore we know that removing those props leaves us
    //    with FCProps
    //
    const linkProps: FCProps = remainingProps as unknown as FCProps
    const classes = linkClasses(variant, className)
    return React.createElement(
      asCustom,
      {
        className: classes,
        ...linkProps,
      },
      children
    )
  } else {
    const { children, className, variant, ...linkProps } = props

    const classes = linkClasses(variant, className)
    return (
      <a className={classes} {...linkProps}>
        {children}
      </a>
    )
  }
}
