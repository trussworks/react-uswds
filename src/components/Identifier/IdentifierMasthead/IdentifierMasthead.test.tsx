import React from 'react'
import { render } from '@testing-library/react'

import dotGovIcon from 'uswds/src/img/icon-dot-gov.svg'

import { IdentifierMasthead } from './IdentifierMasthead'
import { IdentifierLogo } from '../IdentifierLogo/IdentifierLogo'
import { IdentifierLogos } from '../IdentifierLogos/IdentifierLogos'
import { IdentifierIdentity } from '../IdentifierIdentity/IdentifierIdentity'
import { Link } from '../../Link/Link'

const testIdentifierLogo = [
  <img
    key="three"
    src={dotGovIcon}
    className="usa-identifier__logo-img"
    alt="Test Agency Name logo"
  />,
]

const testIdentifierLogoSpanish = [
  <img
    key="four"
    src={dotGovIcon}
    className="usa-identifier__logo-img"
    alt="Logo de Test Agency Name"
  />,
]

describe('IdentifierMasthead component', () => {
  it('renders without errors', () => {
    const { getAllByRole, queryByTestId } = render(
      <IdentifierMasthead
        className="usa-identifier__custom-class-name"
        aria-label="Agency identifier">
        <IdentifierLogos>
          <IdentifierLogo href="#" className="custom-class-name">
            {testIdentifierLogo}
          </IdentifierLogo>
        </IdentifierLogos>
        <IdentifierIdentity domain="domain.edu.mil.gov">
          {`An official website of the `}
          <Link href="#">Test Agency Name</Link>
        </IdentifierIdentity>
      </IdentifierMasthead>
    )

    expect(queryByTestId('identifierMasthead-logo')).toBeInTheDocument
    expect(getAllByRole('link')).toHaveLength(2)
    expect(queryByTestId('identifierMasthead')).toBeInTheDocument()
    expect(queryByTestId('identifierMasthead')).toHaveClass(
      'usa-identifier__section usa-identifier__section--masthead usa-identifier__custom-class-name'
    )
  })

  it('renders in Spanish', () => {
    const { queryByAltText } = render(
      <IdentifierMasthead>{testIdentifierLogoSpanish}</IdentifierMasthead>
    )
    expect(queryByAltText('Logo de Test Agency Name')).toBeInTheDocument()
  })

  it('renders without a logo passed in', () => {
    const { queryByTestId } = render(<IdentifierMasthead />)
    expect(queryByTestId('identifierMasthead-logo')).not.toBeInTheDocument()
  })

  it('renders with more than two logos passed in', () => {
    const { getAllByRole, queryByText } = render(
      <IdentifierMasthead aria-label="Agency identifier">
        <IdentifierLogos>
          <IdentifierLogo href="#">{testIdentifierLogo}</IdentifierLogo>
          <IdentifierLogo href="#">{testIdentifierLogo}</IdentifierLogo>
          <IdentifierLogo href="#">{testIdentifierLogo}</IdentifierLogo>
        </IdentifierLogos>
        <IdentifierIdentity domain="domain.edu.mil.gov">
          {`An official website of the `}
          <Link href="#">Test Agency Name</Link>
          {`, `}
          <Link href="#">Second Test Agency Name</Link>
          {`, and the `}
          <Link href="#">Third Test Agency Name</Link>
        </IdentifierIdentity>
      </IdentifierMasthead>
    )
    expect(getAllByRole('img')).toHaveLength(3)
    expect(getAllByRole('link')).toHaveLength(6)
    expect(queryByText('Third Test Agency Name')).toBeInTheDocument()
  })
})
