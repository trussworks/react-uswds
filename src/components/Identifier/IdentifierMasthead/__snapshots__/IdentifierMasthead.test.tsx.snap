// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentifierMasthead component renders consistently in Spanish when passed Spanish for language prop 1`] = `
<section
  aria-label="Identificador de la agencia"
  className="usa-identifier__section usa-identifier__section--masthead"
  data-testid="identifierMasthead"
>
  <div
    className="usa-identifier__container"
  >
    <div
      className="usa-identifier__logos"
    >
      <a
        className="usa-identifier__logo"
        href="www.google.com"
      >
        <img
          alt="Test Parent Agency Name logo"
          className="usa-identifier__logo-img"
          data-testid="identifierMasthead-logo"
          src="test-file-stub"
        />
      </a>
    </div>
    <div
      aria-label="Descripción de la agencia"
      className="usa-identifier__identity"
      data-testid="identifierMasthead-agency-description"
    >
      <p
        className="usa-identifier__identity-domain"
      >
        aGovDomain.gov
      </p>
      <p
        className="usa-identifier__identity-disclaimer"
      >
        Un sitio web oficial de
         
        <a
          href="https://www.parentAgencyExampleUrl.gov/"
        >
          Test Parent Agency Name
        </a>
      </p>
    </div>
  </div>
</section>
`;
